packages/perspective-viewer-d3fc/src/js/charts/heatmap.js

```diff
@@ -0,0 +1,76 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2017, the Perspective Authors.
+ *
+ * This file is part of the Perspective library, distributed under the terms of
+ * the Apache License 2.0.  The full license can be found in the LICENSE file.
+ *
+ */
+import * as d3 from "d3";
+import * as fc from "d3fc";
+import * as crossAxis from "../axis/crossAxis";
+//import * as otherAxis from "../axis/otherAxis";
+import {heatmapSeries} from "../series/heatmapSeries";
+import {heatmapData} from "../data/heatmapData";
+import {filterData} from "../legend/legend";
+import {getOrCreateElement} from "../utils/utils";
+import scrollableLegend from "../legend/scrollableLegend";
+import chartSvgCartesian from "../d3fc/chart/svg/cartesian";
+import {withGridLines} from "../gridlines/gridlines";
+
+function heatmapChart(container, settings) {
+    const data = heatmapData(settings, filterData(settings));
+    const colorInterpolate = d3.interpolateViridis;
+
+    const series = heatmapSeries(settings, colorInterpolate);
+
+    const uniqueYDomain = [...new Set(data.map(d => d.mainValue))];
+    const extent = fc.extentLinear().accessors([d => d.colorValue]);
+
+    const colourDomain = extent(data);
+    legend(container, colorInterpolate, colourDomain);
+
+    const chart = chartSvgCartesian(crossAxis.scale(settings), crossAxis.scale(settings, "splitValues"))
+        .xDomain(crossAxis.domain(settings)(settings.data))
+        .yDomain(uniqueYDomain)
+        .yOrient("left")
+        .plotArea(withGridLines(series));
+
+    crossAxis.styleAxis(chart, "x", settings, "crossValues");
+    crossAxis.styleAxis(chart, "y", settings, "splitValues", uniqueYDomain);
```

Comment:
If we're passing in the "splitValues` key to tell it how to setup the axis properties, can we avoid creating `uniqueYDomain` altogether?

Reply:
The uniqueYDomain is passed so that I can get the lengths of the individual fields. There might be a better way, but getting it directly from the settings.data would mean calling the heatmapData function to structure the data correctly and this only applies to the heatmap chart and not the generic crossAxis code.

Reply:
If you change crossAxis.domain so that it's based on the data like for mainAxis, would that solve this problem and also negate the need for a crossAxis.domainFromData function?

Reply:
Possibly, but I am a bit worried about breaking all the other charts that use the current implementation of crossAxis.domain. Can you spare some time tomorrow to go through it with me?

Reply:
@DevAndyLee - I have made an attempt at updating the crossAxis.domain to function to reference data rather than settings.data